From e643b0cd96071c430b83ce8c91b7c2c5cf31bd88 Mon Sep 17 00:00:00 2001
From: yzhu1 <yzhu1@windriver.com>
Date: Tue, 24 Nov 2015 17:13:53 +0800
Subject: [PATCH 1/3] Revert "random: use the architectural HWRNG for the
 SHA's IV in extract_buf()"

The backport of the commit 8f494e8 "random: use the architectural HWRNG for
the SHA's IV in extract_buf()" was incomplete. And we are reverting it
to start over with a new set of changes.

Signed-off-by: yzhu1 <yzhu1@windriver.com>
---
 drivers/char/random.c |    6 +++---
 1 files changed, 3 insertions(+), 3 deletions(-)

diff --git a/drivers/char/random.c b/drivers/char/random.c
index 6c03531..b196135 100644
--- a/drivers/char/random.c
+++ b/drivers/char/random.c
@@ -1000,17 +1000,17 @@ static void extract_buf(struct entropy_store *r, __u8 *out)
 
 	/*
 	 * If we have a architectural hardware random number
-	 * generator, use it for SHA's initial vector
+	 * generator, mix that in, too.
 	 */
-	sha_init(hash.w);
 	for (i = 0; i < LONGS(EXTRACT_SIZE); i++) {
 		unsigned long v;
 		if (!arch_get_random_long(&v))
 			break;
-		hash.l[i] = v;
+		hash.l[i] ^= v;
 	}
 
 	/* Generate a hash across the pool, 16 words (512 bits) at a time */
+	sha_init(hash.w);
 	spin_lock_irqsave(&r->lock, flags);
 	for (i = 0; i < r->poolinfo->poolwords; i += 16)
 		sha_transform(hash.w, (__u8 *)(r->pool + i), workspace);
-- 
1.7.5.4

